qiskit/circuit/quantumcircuit.py

```diff
@@ -242,6 +249,26 @@ def data(self, data_input):
         for inst, qargs, cargs in data_input:
             self.append(inst, qargs, cargs)
 
+    @property
+    def metadata(self):
+        """The user provided metadata associated with the circuit
+
+        The metadata for the circuit is a user provided ``dict`` of metadata
+        for the circuit. It will not be used to influence the execution or
+        operation of the circuit, but it is expected to be passed betweeen
+        all transforms of the circuit (ie transpilation) and that providers will
+        associate any circuit metadata with the results it returns from
+        execution of that circuit.
+        """
+        return self._metadata
+
+    @metadata.setter
+    def metadata(self, metadata):
+        """Update the circuit metadata"""
+        if not isinstance(metadata, dict) and metadata is not None:
+            raise TypeError("Only a dictionary or None is accepted for circuit metadata")
+        self._metadata = metadata
```

Comment:
Do we want to provide a way to incrementally add metadata to a circuit by inserting a key, val pair into the dict?

Reply:
Nevermind, I just realized the property attribute doesn't prevent one from adding data via `circ.metadata[key] = val`

Reply:
You can do all of this already by operating on `circuit.metadata` directly. For example. `circuit.metadata['foo'] = 'bar'` will work fine because it's modifying the dictionary returned by the getter above. The setter is only used when you do a `circuit.metadata = foo` call to set the attribute to something